src/wms/src/main/java/org/geoserver/wms/clip/ClipWMSGetMapCallBack.java

```diff
@@ -0,0 +1,152 @@
+                fl.getUserData().putAll(layer.getUserData());
+                return clippedLayer;
+
+            } else if (layer instanceof GridReaderLayer) {
+
+                GridReaderLayer gr = (GridReaderLayer) layer;
+                // wrap
+                CroppedGridCoverage2DReader croppedGridReader =
+                        new CroppedGridCoverage2DReader(gr.getReader(), wktGeom);
+                GridReaderLayer croppedGridLayer =
+                        new GridReaderLayer(croppedGridReader, layer.getStyle());
+                BeanUtilsBean2.getInstance().copyProperties(croppedGridLayer, gr);
+                croppedGridLayer.getUserData().putAll(layer.getUserData());
+                return croppedGridLayer;
+            }
+        } catch (Exception e) {
+            LOGGER.severe("Error occurred while clipping layer " + layer.getTitle());
+            LOGGER.log(Level.SEVERE, e.getMessage(), e);
+            return layer;
+        }
+
+        return layer;
+    }
+
+    @Override
+    public WMSMapContent beforeRender(WMSMapContent mapContent) {
+        return mapContent;
+    }
+
+    @Override
+    public WebMap finished(WebMap map) {
+        return map;
+    }
+
+    @Override
+    public void failed(Throwable t) {}
+
+    private static synchronized Geometry readGeometry(
+            final String wkt, final CoordinateReferenceSystem mapCRS) throws Exception {
+        String[] wktContents = wkt.split(";");
+        Geometry geom = reader.read(wktContents[wktContents.length - 1]);
+        // geom.setSRID(CRS.lookupEpsgCode(mapCRS, false));
+        if (!(geom.getClass().isAssignableFrom(Polygon.class)
+                || geom.getClass().isAssignableFrom(MultiPolygon.class)))
+            throw new ServiceException(
+                    "Clip must be a polygon or multipolygon", "InvalidParameterValue", "clip");
+        // parse SRID if passed
+        // looking for a pattern srid=4326:Polygon(...)
+        if (wktContents.length == 2 && SRID_REGEX.matcher(wktContents[0].toUpperCase()).matches()) {
+            String sridString = wktContents[0].split("=")[1];
+            // force xy
+            CoordinateReferenceSystem geomCRS = CRS.decode("EPSG:" + sridString, true);
+            CoordinateReferenceSystem mapCRSXY =
+                    CRS.decode("EPSG:" + CRS.lookupEpsgCode(mapCRS, false), true);
+            if (CRS.isTransformationRequired(mapCRSXY, geomCRS)) {
+                MathTransform transform = CRS.findMathTransform(geomCRS, mapCRSXY);
+                geom = JTS.transform(geom, transform);
+            }
+        }
+        // finally assign map crs
+        geom.setSRID(CRS.lookupEpsgCode(mapCRS, false));
+        return geom;
+    }
+
+    public static Geometry getClipGeometry(GetMapRequest getMapRequest) {
+
+        // no raw kvp or request has no crs
+        if (getMapRequest.getRawKvp() == null || getMapRequest.getCrs() == null) return null;

```

Comment:
Why go and re-parse the geometry? The GetMapRequest already contains it as a Geometry by now no?

Reply:
Also, if the above is true, this method could become a private one in the KVP reader.

Reply:
will look into this

Reply:
this method itself is used in  `org.geoserver.wms.map.GetMapKvpRequestReader.read(....` , but rightfully pointed out, I will use the getter in `GetMapRequest`